src/librustc_mir/const_eval.rs

```diff
@@ -96,22 +96,25 @@ pub fn op_to_const<'tcx>(
             // FIXME shouldn't it be the case that `mark_static_initialized` has already
             // interned this?  I thought that is the entire point of that `FinishStatic` stuff?
             let alloc = ecx.tcx.intern_const_alloc(alloc);
-            ConstValue::ByRef(ptr.alloc_id, alloc, ptr.offset)
+            (ConstValue::ByRef, Some((alloc, ptr)))
         },
         Ok(Immediate::Scalar(x)) =>
-            ConstValue::Scalar(x.not_undef()?),
+            (ConstValue::Scalar(x.not_undef()?), None),
         Ok(Immediate::ScalarPair(a, b)) =>
-            ConstValue::Slice(a.not_undef()?, b.to_usize(ecx)?),
+            (ConstValue::Slice(a.not_undef()?, b.to_usize(ecx)?), None),
     };
-    Ok(ty::Const { val, ty: op.layout.ty })
+    Ok(ty::Const { val, ty: op.layout.ty, alloc })
 }
```

Comment:
Where is `op_to_const` even still needed? Isn't the purpose of this patch to get rid of such awful conversion functions?

Reply:
This is the groundwork for that. Getting rid of the conversion functions needs some more refactorings in the match code.

Reply:
Oh I see.

What I do not like is that this PR now duplicates the logic about what gets "cached" in `val`. Can you avoid that?

Reply:
But, actually, why can't `const_field` create an `MPlace` from the `alloc` stuff, and thus entirely avoid `lazy_const_to_op` and `op_to_const`?

Reply:
It can. I just really dislike doing many such refactorings at once.

I'll try to avoid duplicating the caching logic

Reply:
deduplicating did not succeed, I will try again once all the `Op` conversion stuff is simplified ot `MemPlace`

Reply:
> I just really dislike doing many such refactorings at once.

Well, that is a completely local change in `const_field`, is it not?  And it proves the viability of the refactoring, so it seems like a good sanity check.

Reply:
If this is *not* just a local change, I am fine doing this in separate PRs, but then I'd like to understand why that is the case. :)

Reply:
I just tried it out.... I can indeed avoid `op_to_const` completely, but `lazy_const_to_op` is still necessary for `const_discriminant`, because `read_discriminant` takes an `OpTy`.

Reply:
Well, that's a start. :)

Still, I am confused. I assume you have some kind of `const_to_mplace` for this? It can take a `ty::Const`, that wraps up a `Ty` and the allocation stuff so we can compute the layout, which is great. Once you have an `MPlaceTy`, you can use all the great projection functions, and you can call `into()` to get an `OpTy` for `read_discriminant`.

Reply:
While we can do that, it will mean that reading the discriminant from an `Option<&T>` will need to go through the memory instead of just scraping it from the value directly.

Oh... that happens for the field access, too. Well.. we can just bench it and see how it goes

Reply:
Hm, yes, that's true.

But I just realized that maybe we can have a `lazy_const_to_op` that's not quite as bad because it doesn't take a `ty`? The `Evaluated` case contains a `Ty`, after all, and the `Unevaluated` case should have all the information to get it. Or can we even exclude `Unevaluated` entirely because we only use this in pattern matching on already evaluated stuff? `const_to_op` (taking `ty::Const` seems fairly reasonable, *if* it does not take a `Ty` or `TyLayout`.

Reply:
There's still a use of `lazy_const_to_op` left in the const propagator, and I'm at a total loss how to get rid of that one. I'll implement your suggestion.

Reply:
Hm... looking at that, can't it use `ecx.eval_operand(Operand::Constant(Box::new(c)))`?